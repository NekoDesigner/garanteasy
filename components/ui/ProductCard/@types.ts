export interface IProductCardProps {
  style?: object;
  testID?: string;
  image?: string | { uri: string } | number; // string for file URIs, {uri: string} for remote URIs, number for require()
  brand: string;
  name: string;
  purchaseDate: Date | string;
  warrantyDuration: string;
}