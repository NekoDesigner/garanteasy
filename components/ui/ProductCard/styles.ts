import { StyleSheet } from 'react-native';
import { COLORS, SIZES } from '../../../constants';

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-start',
    padding: 8,
    backgroundColor: COLORS.greyLighter,
   borderColor: COLORS.greyDarker,
    borderWidth: 1,
    borderRadius: 8,
    gap: 8,
    maxWidth: 327,
  },
  image: {
    borderRadius: 4,
    width: 64,
    height: 64,
  },
  tags: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    gap: 8,
    marginTop: 8,
    marginBottom: 8,
  },
  tagsText: {
    color: COLORS.blueDarker,
    fontFamily: SIZES.font.familly.default,
    fontSize: SIZES.font.xxs,
  },
  brandContainer: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    marginBottom: 4,
},
  name: {
    color: COLORS.blueDarker,
    textOverflow: 'ellipsis',
    overflow: 'hidden',
    fontFamily: SIZES.font.familly.default,
    fontWeight: SIZES.font.weight.semiBold,
    fontSize: SIZES.font.xl,
  },
  brand: {
    color: COLORS.blueDarker,
    textOverflow: 'ellipsis',
    overflow: 'hidden',
    fontFamily: SIZES.font.familly.default,
    fontWeight: SIZES.font.weight.semiBold,
    fontSize: SIZES.font.xs,
  },
});

export default styles;
