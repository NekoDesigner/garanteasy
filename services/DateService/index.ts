export class DateService {
  static IsDateAfter(date: Date, referenceDate: Date): boolean {
    const dateToCheck = new Date(date);
    const refDate = new Date(referenceDate);
    return dateToCheck.getTime() > refDate.getTime();
  }
  static getDaysBetweenDates(startDate: Date, endDate: Date): number {
    const start = new Date(startDate);
    const end = new Date(endDate);
    const timeDiff = end.getTime() - start.getTime();
    return Math.ceil(timeDiff / (1000 * 3600 * 24));
  }
  static addDays(date: Date, days: number): Date {
    const result = new Date(date);
    result.setDate(result.getDate() + days);
    return result;
  }

  static subtractDays(date: Date, days: number): Date {
    const result = new Date(date);
    result.setDate(result.getDate() - days);
    return result;
  }

  static formatDate(date: Date, locale: string = 'fr-FR'): string {
    return date.toLocaleDateString(locale);
  }

  static formatDDMM(date: Date): string {
    const day = date.getDate().toString().padStart(2, '0');
    const month = (date.getMonth() + 1).toString().padStart(2, '0'); // Months are zero-based
    return `${day}/${month}`;
  }

  static formatMMYY(date: Date | string): string {
    if (typeof date === 'string') {
      date = new Date(date);
    }
    const month = (date.getMonth() + 1).toString().padStart(2, '0'); // Months are zero-based
    const year = date.getFullYear().toString().slice(-2); // Get last two digits of the year
    return `${month}/${year}`;
  }
}

