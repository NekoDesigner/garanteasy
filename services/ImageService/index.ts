import * as FileSystem from 'expo-file-system/legacy';

export interface ImageSaveOptions {
  format?: 'jpeg' | 'png';
  filename?: string;
  directory?: string;
}

export interface ImageInfo {
  uri: string;
  width?: number;
  height?: number;
  size?: number;
}

/**
 * ImageService - Handles saving and managing images in the app's filesystem
 */
export class ImageService {
  // Directory for storing item images
  private static readonly ITEMS_IMAGES_DIR = `${FileSystem.documentDirectory}images/items/`;
  private static readonly DOCUMENTS_IMAGES_DIR = `${FileSystem.documentDirectory}images/documents/`;

  /**
   * Initialize the image directories
   */
  static async initializeDirectories(): Promise<void> {
    try {
      // Create items images directory
      const itemsDirInfo = await FileSystem.getInfoAsync(this.ITEMS_IMAGES_DIR);
      if (!itemsDirInfo.exists) {
        await FileSystem.makeDirectoryAsync(this.ITEMS_IMAGES_DIR, { intermediates: true });
      }

      // Create documents images directory
      const docsDirInfo = await FileSystem.getInfoAsync(this.DOCUMENTS_IMAGES_DIR);
      if (!docsDirInfo.exists) {
        await FileSystem.makeDirectoryAsync(this.DOCUMENTS_IMAGES_DIR, { intermediates: true });
      }
    } catch (error) {
      console.error('Error initializing image directories:', error);
      throw new Error(`Failed to initialize image directories: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Save an item image to the filesystem
   * @param imageUri The URI of the image to save
   * @param itemId Optional item ID for naming
   * @param options Save options
   * @returns Promise<string> The new file URI
   */
  static async saveItemImage(
    imageUri: string,
    itemId?: string,
    options: ImageSaveOptions = {}
  ): Promise<string> {
    try {
      await this.initializeDirectories();

      const {
        format = 'jpeg',
        filename,
      } = options;

      // Generate unique filename
      const timestamp = Date.now();
      const randomId = Math.random().toString(36).substring(2, 15);
      const extension = format === 'jpeg' ? 'jpg' : format;

      const finalFilename = filename ||
        (itemId ? `item-${itemId}-${timestamp}.${extension}` : `item-${randomId}-${timestamp}.${extension}`);

      const targetPath = `${this.ITEMS_IMAGES_DIR}${finalFilename}`;

      // Copy the image to the target location
      await FileSystem.copyAsync({
        from: imageUri,
        to: targetPath,
      });

      return targetPath;
    } catch (error) {
      console.error('Error saving item image:', error);
      throw new Error(`Failed to save item image: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Save a document image to the filesystem
   * @param imageUri The URI of the image to save
   * @param documentId Optional document ID for naming
   * @param options Save options
   * @returns Promise<string> The new file URI
   */
  static async saveDocumentImage(
    imageUri: string,
    documentId?: string,
    options: ImageSaveOptions = {}
  ): Promise<string> {
    try {
      await this.initializeDirectories();

      const {
        format = 'jpeg',
        filename,
      } = options;

      // Generate unique filename
      const timestamp = Date.now();
      const randomId = Math.random().toString(36).substring(2, 15);
      const extension = format === 'jpeg' ? 'jpg' : format;

      const finalFilename = filename ||
        (documentId ? `doc-${documentId}-${timestamp}.${extension}` : `doc-${randomId}-${timestamp}.${extension}`);

      const targetPath = `${this.DOCUMENTS_IMAGES_DIR}${finalFilename}`;

      // Copy the image to the target location
      await FileSystem.copyAsync({
        from: imageUri,
        to: targetPath,
      });

      return targetPath;
    } catch (error) {
      console.error('Error saving document image:', error);
      throw new Error(`Failed to save document image: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Get image information
   * @param imageUri The URI of the image
   * @returns Promise<ImageInfo>
   */
  static async getImageInfo(imageUri: string): Promise<ImageInfo> {
    try {
      const fileInfo = await FileSystem.getInfoAsync(imageUri);

      return {
        uri: imageUri,
        size: fileInfo.exists ? (fileInfo as any).size : undefined,
        // Note: Getting image dimensions would require additional libraries like expo-image-manipulator
        // For now, we'll just return the URI and size
      };
    } catch (error) {
      console.error('Error getting image info:', error);
      throw new Error(`Failed to get image info: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Delete an image file
   * @param imageUri The URI of the image to delete
   * @returns Promise<void>
   */
  static async deleteImage(imageUri: string): Promise<void> {
    try {
      const fileInfo = await FileSystem.getInfoAsync(imageUri);
      if (fileInfo.exists) {
        await FileSystem.deleteAsync(imageUri);
      }
    } catch (error) {
      console.error('Error deleting image:', error);
      throw new Error(`Failed to delete image: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Check if an image URI is a local file URI (starts with file://)
   * @param uri The URI to check
   * @returns boolean
   */
  static isLocalFileUri(uri: string): boolean {
    return uri.startsWith('file://') || uri.startsWith(FileSystem.documentDirectory || '');
  }

  /**
   * Check if an image file exists
   * @param imageUri The URI of the image
   * @returns Promise<boolean>
   */
  static async imageExists(imageUri: string): Promise<boolean> {
    try {
      const fileInfo = await FileSystem.getInfoAsync(imageUri);
      return fileInfo.exists;
    } catch {
      return false;
    }
  }

  /**
   * List all item images
   * @returns Promise<string[]>
   */
  static async listItemImages(): Promise<string[]> {
    try {
      await this.initializeDirectories();

      const files = await FileSystem.readDirectoryAsync(this.ITEMS_IMAGES_DIR);
      return files
        .filter(file => file.toLowerCase().match(/\.(jpg|jpeg|png)$/))
        .map(file => `${this.ITEMS_IMAGES_DIR}${file}`);
    } catch (error) {
      console.error('Error listing item images:', error);
      return [];
    }
  }

  /**
   * List all document images
   * @returns Promise<string[]>
   */
  static async listDocumentImages(): Promise<string[]> {
    try {
      await this.initializeDirectories();

      const files = await FileSystem.readDirectoryAsync(this.DOCUMENTS_IMAGES_DIR);
      return files
        .filter(file => file.toLowerCase().match(/\.(jpg|jpeg|png)$/))
        .map(file => `${this.DOCUMENTS_IMAGES_DIR}${file}`);
    } catch (error) {
      console.error('Error listing document images:', error);
      return [];
    }
  }

  /**
   * Clean up orphaned images (images that don't belong to any existing items or documents)
   * This is a utility function for maintenance
   * @param existingItemIds Array of existing item IDs
   * @param existingDocumentIds Array of existing document IDs
   * @returns Promise<number> Number of images deleted
   */
  static async cleanupOrphanedImages(
    existingItemIds: string[] = [],
    existingDocumentIds: string[] = []
  ): Promise<number> {
    try {
      let deletedCount = 0;

      // Clean item images
      const itemImages = await this.listItemImages();
      for (const imagePath of itemImages) {
        const filename = imagePath.split('/').pop() || '';
        const hasValidItemId = existingItemIds.some(id => filename.includes(`item-${id}-`));

        if (!hasValidItemId) {
          await this.deleteImage(imagePath);
          deletedCount++;
        }
      }

      // Clean document images
      const docImages = await this.listDocumentImages();
      for (const imagePath of docImages) {
        const filename = imagePath.split('/').pop() || '';
        const hasValidDocId = existingDocumentIds.some(id => filename.includes(`doc-${id}-`));

        if (!hasValidDocId) {
          await this.deleteImage(imagePath);
          deletedCount++;
        }
      }

      return deletedCount;
    } catch (error) {
      console.error('Error cleaning up orphaned images:', error);
      throw new Error(`Failed to cleanup orphaned images: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }
}
