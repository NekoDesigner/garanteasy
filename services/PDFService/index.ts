import 'react-native-get-random-values'; // Required for pdf-lib
import { Buffer } from 'buffer';
import * as FileSystem from 'expo-file-system/legacy';
import * as Sharing from 'expo-sharing';
import { PDFDocument } from 'pdf-lib';

export interface PDFCreationOptions {
  title?: string;
  author?: string;
  subject?: string;
  compress?: boolean;
  pageMargin?: number;
  imageQuality?: number;
  useTimestamp?: boolean; // Whether to use timestamp in filename
}

export interface ImageInfo {
  uri: string;
  width?: number;
  height?: number;
}

/**
 * Create a PDF document from an array of image URIs
 * @param imageUris Array of image URIs to convert to PDF
 * @param options Optional PDF creation settings
 * @returns Promise<string> The file path of the created PDF
 */
export const createPdfFromImages = async (
  imageUris: string[],
  options: PDFCreationOptions = {}
): Promise<string> => {
  if (!imageUris || imageUris.length === 0) {
    throw new Error('No images provided for PDF creation');
  }

  const {
    title = 'Scanned Document',
    author = 'GarantEasy Scanner',
    subject = 'Scanned Document',
    compress = true,
    pageMargin = 20,
    useTimestamp = false,
  } = options;

  try {
    // Create a new PDF document
    const pdfDoc = await PDFDocument.create();

    // Set document metadata
    pdfDoc.setTitle(title);
    pdfDoc.setAuthor(author);
    pdfDoc.setSubject(subject);
    pdfDoc.setCreator('GaranteAsy PDF Service');
    pdfDoc.setProducer('pdf-lib');
    pdfDoc.setCreationDate(new Date());
    pdfDoc.setModificationDate(new Date());

    // Process each image
    for (let i = 0; i < imageUris.length; i++) {
      const imageUri = imageUris[i];

      try {
        // Read the image file
        const imageBytes = await FileSystem.readAsStringAsync(imageUri, {
          encoding: FileSystem.EncodingType.Base64,
        });

        // Convert base64 to bytes
        const imageData = `data:image/jpeg;base64,${imageBytes}`;

        // Embed the image in the PDF
        let image;
        if (imageUri.toLowerCase().includes('.png')) {
          image = await pdfDoc.embedPng(imageData);
        } else {
          // Default to JPEG for most scanned documents
          image = await pdfDoc.embedJpg(imageData);
        }

        // Get image dimensions
        const { width: imgWidth, height: imgHeight } = image;

        // Create a new page with appropriate size
        // Calculate page size based on image aspect ratio (A4 size as default)
        const maxWidth = 595; // A4 width in points
        const maxHeight = 842; // A4 height in points

        let pageWidth = maxWidth;
        let pageHeight = maxHeight;

        // Calculate scaling to fit image on page while maintaining aspect ratio
        const aspectRatio = imgWidth / imgHeight;

        if (aspectRatio > 1) {
          // Landscape orientation
          pageHeight = pageWidth / aspectRatio;
          if (pageHeight > maxHeight) {
            pageHeight = maxHeight;
            pageWidth = pageHeight * aspectRatio;
          }
        } else {
          // Portrait orientation
          pageWidth = pageHeight * aspectRatio;
          if (pageWidth > maxWidth) {
            pageWidth = maxWidth;
            pageHeight = pageWidth / aspectRatio;
          }
        }

        const page = pdfDoc.addPage([pageWidth, pageHeight]);

        // Calculate image size and position on page
        const availableWidth = pageWidth - (pageMargin * 2);
        const availableHeight = pageHeight - (pageMargin * 2);

        let finalWidth = availableWidth;
        let finalHeight = availableHeight;

        // Maintain aspect ratio while fitting within available space
        if (aspectRatio > availableWidth / availableHeight) {
          finalHeight = finalWidth / aspectRatio;
        } else {
          finalWidth = finalHeight * aspectRatio;
        }

        // Center the image on the page
        const x = (pageWidth - finalWidth) / 2;
        const y = (pageHeight - finalHeight) / 2;

        // Draw the image on the page
        page.drawImage(image, {
          x,
          y,
          width: finalWidth,
          height: finalHeight,
        });

      } catch (imageError) {
        console.error(`Error processing image ${i + 1}:`, imageError);
        // Continue with other images even if one fails
        continue;
      }
    }

    // Serialize the PDF document
    const pdfBytes = await pdfDoc.save({
      useObjectStreams: compress,
    });

    // Generate unique filename with timestamp
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const filename = title ? `${title}${useTimestamp ? ('-' + timestamp) : ''}.pdf` : `scanned-document-${timestamp}.pdf`;
    const filePath = `${FileSystem.documentDirectory}${filename}`;

    // Write PDF to file system
    // Convert Uint8Array to base64 string using Buffer polyfill
    const base64String = Buffer.from(pdfBytes).toString('base64');
    await FileSystem.writeAsStringAsync(
      filePath,
      base64String,
      { encoding: FileSystem.EncodingType.Base64 }
    );

    return filePath;

  } catch (error) {
    console.error('Error creating PDF:', error);
    throw new Error(`Failed to create PDF: ${error instanceof Error ? error.message : 'Unknown error'}`);
  }
};

/**
 * Create a PDF and immediately share it
 * @param imageUris Array of image URIs to convert to PDF
 * @param options Optional PDF creation settings
 * @returns Promise<void>
 */
export const createAndSharePdf = async (
  imageUris: string[],
  options: PDFCreationOptions = {}
): Promise<void> => {
  try {
    const pdfPath = await createPdfFromImages(imageUris, options);

    // Check if sharing is available
    if (await Sharing.isAvailableAsync()) {
      await Sharing.shareAsync(pdfPath, {
        mimeType: 'application/pdf',
        dialogTitle: 'Share Scanned Document',
      });
    } else {
      throw new Error('Sharing is not available on this device');
    }
  } catch (error) {
    console.error('Error creating and sharing PDF:', error);
    throw error;
  }
};

/**
 * Get file info for a PDF file
 * @param filePath Path to the PDF file
 * @returns Promise<FileSystem.FileInfo>
 */
export const getPdfInfo = async (filePath: string): Promise<FileSystem.FileInfo> => {
  try {
    return await FileSystem.getInfoAsync(filePath);
  } catch (error) {
    console.error('Error getting PDF info:', error);
    throw new Error(`Failed to get PDF info: ${error instanceof Error ? error.message : 'Unknown error'}`);
  }
};

/**
 * Delete a PDF file
 * @param filePath Path to the PDF file to delete
 * @returns Promise<void>
 */
export const deletePdf = async (filePath: string): Promise<void> => {
  try {
    await FileSystem.deleteAsync(filePath);
  } catch (error) {
    console.error('Error deleting PDF:', error);
    throw new Error(`Failed to delete PDF: ${error instanceof Error ? error.message : 'Unknown error'}`);
  }
};

/**
 * List all PDF files in the documents directory
 * @returns Promise<string[]> Array of PDF file paths
 */
export const listPdfFiles = async (): Promise<string[]> => {
  try {
    const files = await FileSystem.readDirectoryAsync(FileSystem.documentDirectory || '');
    return files
      .filter(file => file.toLowerCase().endsWith('.pdf'))
      .map(file => `${FileSystem.documentDirectory}${file}`);
  } catch (error) {
    console.error('Error listing PDF files:', error);
    throw new Error(`Failed to list PDF files: ${error instanceof Error ? error.message : 'Unknown error'}`);
  }
};

export const removeFileExtension = (filename: string): string => {
  const lastDotIndex = filename.lastIndexOf('.');
  if (lastDotIndex === -1) return filename; // No extension found
  return filename.substring(0, lastDotIndex);
};