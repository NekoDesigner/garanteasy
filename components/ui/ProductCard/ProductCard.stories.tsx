/**
 * ProductCard - Generated by Garanteasier CLI
 */

import React from "react";
import { View } from "react-native";
import { DateService } from "../../../services/DateService";
import ProductCard from "./index";

export default {
  title: 'ProductCard',
  component: ProductCard
};

export const Default = () => (
    <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center', padding: 8, gap: 16 }}>
    <ProductCard
      brand="Bosh"
      name="Tondeuse à gazon"
      purchaseDate={DateService.subtractDays(new Date(), 3)}
      warrantyDuration="15d"
      image={require('../../../assets/images/tondeuse-test.png')}
    />
    <ProductCard
      brand="Qilive"
      name="Four"
      purchaseDate={DateService.subtractDays(new Date(), 3)}
      warrantyDuration="2m7j"
      image={require('../../../assets/images/four-test.png')}
    />
    <ProductCard
      brand="Whirpool"
      name="Machine à laver"
      purchaseDate={DateService.subtractDays(new Date(), 8)}
      warrantyDuration="15d"
      image={require('../../../assets/images/washingmachin-test.png')}
    />
    <ProductCard
      brand="Default Brand"
      name="Default Product Name"
      purchaseDate={DateService.subtractDays(new Date(), 11)}
      warrantyDuration="15d"
      image={require('../../../assets/images/default-product.png')}
    />
    </View>
);