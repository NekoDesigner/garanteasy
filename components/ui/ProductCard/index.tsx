/**
 * ProductCard - Generated by Garanteasier CLI
 */

import React from "react";
import { View, Text, Image } from "react-native";
import { DateService } from "../../../services/DateService";
import ProgressIndicator from "../ProgressIndicator";
import RoundedIconButton from "../RoundedIconButton";
import Tag from "../Tag";
import { IProductCardProps } from "./@types";
import ProductCardStyles from "./styles";

const ProductCard: React.FC<IProductCardProps> = ({ style, testID = 'productcard', purchaseDate, ...props }) => {

  const [warrantyDurationTag, setWarrantyDurationTag] = React.useState<string>('');
  const [progress, setProgress] = React.useState<number>(1);

  function getWarrantyDurationInDay(warrantyDurationStr: string): number {
    let result = 0;
    warrantyDurationStr = warrantyDurationStr.toLowerCase();
    const warrantyDurationInYear = warrantyDurationStr.split('y')[0].trim();
    const hasYear = warrantyDurationStr.includes('y');
    if (warrantyDurationInYear && hasYear) {
      const warrantyDurationInDays = parseInt(warrantyDurationInYear) * 365;
      result += warrantyDurationInDays;
    }
    const warrantyDurationInMonth = warrantyDurationStr.split('m')[0].trim();
    const hasMonth = warrantyDurationStr.includes('m');
    if (warrantyDurationInMonth && hasMonth) {
      const warrantyDurationInDays = parseInt(warrantyDurationInMonth) * 30;
      result += warrantyDurationInDays;
    }
    const warrantyDurationInDays = warrantyDurationStr.split('d')[0].trim();
    const hasDays = warrantyDurationStr.includes('d');
    if (warrantyDurationInDays && hasDays) {
      result += parseInt(warrantyDurationInDays);
    }
    return result;
  }

  function formatExpirationDate(purchaseDateProp: Date | string, warrantyDurationInDays: number): string {
    const _purchaseDate = new Date(purchaseDateProp);
    const expirationDate = DateService.addDays(_purchaseDate, warrantyDurationInDays);
    const currentDate = new Date(); // Use the current date for calculations

    if (DateService.IsDateAfter(currentDate, expirationDate)) {
      return `Expiré le ${DateService.formatMMYY(expirationDate)}`;
    }

    // Calculate days left from the current date to the expiration date
    const daysLeft = DateService.getDaysBetweenDates(currentDate, expirationDate);

    if (daysLeft > 365) {
      const yearsLeft = Math.floor(daysLeft / 365);
      return `Expire dans ${yearsLeft}an${yearsLeft > 1 ? 's' : ''}`;
    }
    if (daysLeft > 99) {
      const monthsLeft = Math.floor(daysLeft / 30);
      return `Expire dans ${monthsLeft}m`;
    }
    return `Expire dans ${daysLeft}j`;
  }

  React.useEffect(() => {
    if (props.warrantyDuration) {
      const warrantyDurationInDays = getWarrantyDurationInDay(props.warrantyDuration);
      const expirationDate = formatExpirationDate(purchaseDate, warrantyDurationInDays);
      setWarrantyDurationTag(expirationDate);

      // Calculate rest days from purchase date to expiration date
      const _purchaseDate = new Date(purchaseDate);
      const expirationDateObj = DateService.addDays(_purchaseDate, warrantyDurationInDays);
      const currentDate = new Date();
      if (DateService.IsDateAfter(currentDate, expirationDateObj)) {
        setProgress(0);
      }
      else if (DateService.IsDateAfter(currentDate, _purchaseDate)) {
        const totalDays = DateService.getDaysBetweenDates(_purchaseDate, expirationDateObj);
        const daysPassed = DateService.getDaysBetweenDates(_purchaseDate, currentDate);

        setProgress(1 - (daysPassed / totalDays));
      }
      else {
        setProgress(1);
      }

    } else {
      setWarrantyDurationTag('Aucune garantie');
    }
  }, [props.warrantyDuration, purchaseDate]);

  return (
      <View style={[ProductCardStyles.container, style]} testID={testID} {...props}>
        <Image
          source={props.image || require('../../../assets/images/default-product.png')}
          style={ProductCardStyles.image}
          resizeMode="cover"
        />
        <View style={{ display: 'flex', flex: 1 }}>
          <View style={ProductCardStyles.brandContainer}>
            <Text style={ProductCardStyles.name} numberOfLines={1}>
              {props.name || 'Product Name'}
            </Text>
            <RoundedIconButton icon="arrow-right" onPress={() => {}} size="small" />
          </View>
          <Text style={ProductCardStyles.brand} numberOfLines={1}>
            {props.brand || 'Brand Name'}
          </Text>
          <View style={ProductCardStyles.tags}>
            <Tag testID="test-tag-1" label={`Acheté le: ${DateService.formatMMYY(purchaseDate)}`} textStyle={ProductCardStyles.tagsText} />
            <Tag testID="test-tag-2" label={warrantyDurationTag} textStyle={ProductCardStyles.tagsText} />
        </View>
        <ProgressIndicator value={progress} />
        </View>
      </View>
  );
};

export default ProductCard;
