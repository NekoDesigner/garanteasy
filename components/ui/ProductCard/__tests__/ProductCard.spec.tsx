/**
 * ProductCard - Generated by Garanteasier CLI
 */

import { render, waitFor } from "@testing-library/react-native";
import React from "react";
import { DateService } from "../../../../services/DateService";
import ProductCard from "../index";

describe("ProductCard Components", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

    test('renders correctly', () => {
      const tree = render(<ProductCard
        brand="Bosh"
        name="Tondeuse à gazon"
        purchaseDate={DateService.subtractDays(new Date(), 3)}
        warrantyDuration="15d"
        image={require('../../../../assets/images/default-product.png')}
      />).toJSON();
      expect(tree).toMatchSnapshot();
    });

  beforeEach(() => {
    jest.useFakeTimers().setSystemTime(new Date()); // Mock the current date
  });

  afterEach(() => {
    jest.useRealTimers(); // Restore real timers
  });

  describe('warranty duration progress indicator', () => {
    it('should have full progress indicator bar', async () => {
        const { getByTestId } = render(
          <ProductCard
            brand="Bosh"
            name="Tondeuse à gazon"
            purchaseDate={new Date()}
            warrantyDuration="15d"
            image={require('../../../../assets/images/default-product.png')}
          />
        );

        const progressBar = getByTestId('progressindicator');
        expect(progressBar).toBeTruthy();

        await waitFor(() => {
          const progressBarStyle = progressBar.props.style[0]; // Access the first style object
          expect(progressBarStyle.width).toBe("100%"); // Adjust the expected width based on your implementation
        });
    });

    it('should have full progress indicator bar', async () => {
        const { getByTestId } = render(
          <ProductCard
            brand="Bosh"
            name="Tondeuse à gazon"
            purchaseDate={DateService.subtractDays(new Date(), 3)}
            warrantyDuration="15d"
            image={require('../../../../assets/images/default-product.png')}
          />
        );

        const progressBar = getByTestId('progressindicator');
        expect(progressBar).toBeTruthy();

        await waitFor(() => {
          const progressBarStyle = progressBar.props.style[0]; // Access the first style object
          expect(progressBarStyle.width).not.toBe("100%"); // Adjust the expected width based on your implementation
        });
    });

    it('should have full progress indicator bar', async () => {
        const { getByTestId } = render(
          <ProductCard
            brand="Bosh"
            name="Tondeuse à gazon"
            purchaseDate={DateService.subtractDays(new Date(), 10)}
            warrantyDuration="20d"
            image={require('../../../../assets/images/default-product.png')}
          />
        );

        const progressBar = getByTestId('progressindicator');
        expect(progressBar).toBeTruthy();

        await waitFor(() => {
          const progressBarStyle = progressBar.props.style[0]; // Access the first style object
          expect(progressBarStyle.width).toBe("50%"); // Adjust the expected width based on your implementation
        });
    });
  });

  describe('warranty duration tag', () => {
    it('should display the correct warranty duration tag', async () => {
      const { getByTestId } = render(
        <ProductCard
          brand="Bosh"
          name="Tondeuse à gazon"
          purchaseDate={DateService.subtractDays(new Date(), 3)} // 3 days ago
          warrantyDuration="15d"
          image={require('../../../../assets/images/default-product.png')}
        />
      );

      await waitFor(() => {
        const tag = getByTestId('test-tag-2-text');
        expect(tag).toBeTruthy();
        expect(tag.children).toContain('Expire dans 12j'); // 15 days - 3 days passed
      });
    });

    it('should display "Aucune garantie" when no warranty duration is provided', () => {
      const { getByText } = render(
        <ProductCard
          brand="Bosh"
          name="Tondeuse à gazon"
          purchaseDate={DateService.subtractDays(new Date(), 3)}
          warrantyDuration=""
          image={require('../../../../assets/images/default-product.png')}
        />
      );

      expect(getByText('Aucune garantie')).toBeTruthy();
    });
  });
});