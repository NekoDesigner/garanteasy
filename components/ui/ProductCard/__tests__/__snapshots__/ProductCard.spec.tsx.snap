// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard Components renders correctly 1`] = `
<View
  brand="Bosh"
  image={
    {
      "testUri": "../../../assets/images/default-product.png",
    }
  }
  name="Tondeuse à gazon"
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#F2F2F2",
        "borderColor": "#D9D9D9",
        "borderRadius": 8,
        "borderWidth": 1,
        "display": "flex",
        "flexDirection": "row",
        "gap": 8,
        "justifyContent": "flex-start",
        "padding": 8,
      },
      undefined,
    ]
  }
  testID="productcard"
  warrantyDuration="15d"
>
  <Image
    onError={[Function]}
    resizeMode="cover"
    source={
      {
        "testUri": "../../../assets/images/default-product.png",
      }
    }
    style={
      {
        "borderRadius": 4,
        "height": 64,
        "width": 64,
      }
    }
  />
  <View
    style={
      {
        "display": "flex",
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 4,
        }
      }
    >
      <Text
        numberOfLines={1}
        style={
          {
            "color": "#050426",
            "fontFamily": "Ubuntu",
            "fontSize": 20,
            "fontWeight": "700",
            "overflow": "hidden",
            "textOverflow": "ellipsis",
          }
        }
      >
        Tondeuse à gazon
      </Text>
    </View>
    <Text
      numberOfLines={1}
      style={
        {
          "color": "#050426",
          "fontFamily": "Ubuntu",
          "fontSize": 14,
          "fontWeight": "700",
          "overflow": "hidden",
          "textOverflow": "ellipsis",
        }
      }
    >
      Bosh
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "gap": 8,
          "justifyContent": "space-between",
          "marginBottom": 8,
          "marginTop": 8,
        }
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "backgroundColor": "#FFFFFF",
              "borderRadius": 4,
              "display": "flex",
              "flexDirection": "row",
              "gap": 4,
              "justifyContent": "space-between",
              "paddingHorizontal": 8,
              "paddingVertical": 4,
            },
            {},
          ]
        }
        testID="test-tag-1"
      >
        <Text
          style={
            [
              {
                "color": "#050426",
                "fontFamily": "Ubuntu",
                "fontSize": 14,
                "fontWeight": "700",
              },
              {
                "color": "#050426",
                "fontFamily": "Ubuntu",
                "fontSize": 12,
              },
            ]
          }
          testID="test-tag-1-text"
        >
          Acheté le: 07/25
        </Text>
      </View>
      <View
        style={
          [
            {
              "alignItems": "center",
              "backgroundColor": "#FFFFFF",
              "borderRadius": 4,
              "display": "flex",
              "flexDirection": "row",
              "gap": 4,
              "justifyContent": "space-between",
              "paddingHorizontal": 8,
              "paddingVertical": 4,
            },
            {},
          ]
        }
        testID="test-tag-2"
      >
        <Text
          style={
            [
              {
                "color": "#050426",
                "fontFamily": "Ubuntu",
                "fontSize": 14,
                "fontWeight": "700",
              },
              {
                "color": "#050426",
                "fontFamily": "Ubuntu",
                "fontSize": 12,
              },
            ]
          }
          testID="test-tag-2-text"
        >
          Expire dans 12j
        </Text>
      </View>
    </View>
    <View
      style={
        [
          {
            "backgroundColor": "#7ABA8E",
            "borderRadius": 9999,
            "height": 8,
            "width": "80%",
          },
          undefined,
        ]
      }
      testID="progressindicator"
    />
  </View>
</View>
`;
